include/swift/Parse/Parser.h

```diff
@@ -700,11 +700,54 @@ class Parser {
   bool parseMatchingToken(tok K, SourceLoc &TokLoc, Diag<> ErrorDiag,
                           SourceLoc OtherLoc);
 
+  /// Tracking IfConfigDecl for conditionals in collections for AST dump.
+  /// Since there is no place to directly represent conditionals in the
+  /// current AST collection model, a map is generated from the location
+  /// of actual data values to the conditionals that directly precede them.
+  /// This is transferred to the CollectionExpr when the collection completes
+  /// and is used by the ATSDumper to print a repsentation of the conditional
+  /// as well as by libSyntax and to remove conditionals from module interfaces.
+  /// Nested conditionals are elements of their outer IfConfigDecl structure.
+  class ConfigMap {
+    /// Map accumulated of conditionals preceeding element number.
+    SmallVector<CollectionExpr::ConditionalsMapPair, 2> ConditionalsMap;
+    /// Active element number any #if conditionals will be before.
+    unsigned NextActiveIndexNumber = 0;
+    /// Element number to group conditionals under.
+    unsigned PendingIndexNumber = 0;
+
+  public:
+    Expr *registerActiveDataElement(Expr *expr) {
+      ++NextActiveIndexNumber;
+      return expr;
+    }
+
+    void newOuterConditionalStarts() {
+      PendingIndexNumber = NextActiveIndexNumber;
+    }
+
+    void outerConditionalCompletes(IfConfigDecl *ICD) {
+      ConditionalsMap.emplace_back(std::make_pair(PendingIndexNumber, ICD));
```

Comment:
Could this be just `ConditionalsMap.emplace_back(PendingIndexNumber, ICD)` maybe?

Reply:
I have a lot to learn about c++!